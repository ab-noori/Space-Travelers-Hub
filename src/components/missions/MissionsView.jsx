import React from 'react';
import PropTypes from 'prop-types';
import './MissionsView.scss';

const MissionsView = ({
  missions, isLoading, error, handleJoinMission, handleLeaveMission,
}) => {
  if (isLoading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <div>
      <table className="striped-table">
        <thead>
          <tr>
            <th className="nowrap">Mission Name</th>
            <th className="nowrap">Description</th>
            <th className="nowrap">Status</th>
            <th className="nowrap">Actions</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission, index) => (
            <tr key={mission.mission_id} className={index % 2 === 0 ? 'even-row' : 'odd-row'}>
              <td className="bold">{mission.mission_name}</td>
              <td>{mission.description}</td>
              <td className="button-cell">
                {mission.reserved ? (
                  <button type="button" className="active-member status-btn">
                    Active Member
                  </button>
                ) : (
                  <button type="button" className="not-member status-btn">
                    Not Member
                  </button>
                )}
              </td>
              <td className="button-cell">
                {mission.reserved ? (
                  <button type="button" className="leave-mission action-btn" onClick={() => handleLeaveMission(mission.mission_id)}>
                    Leave Mission
                  </button>
                ) : (
                  <button
                    type="button"
                    className="join-mission action-btn"
                    onClick={() => handleJoinMission(mission.mission_id)}
                  >
                    Join Mission
                  </button>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

MissionsView.propTypes = {
  missions: PropTypes.arrayOf(
    PropTypes.shape({
      mission_id: PropTypes.number.isRequired,
      mission_name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      reserved: PropTypes.bool.isRequired,
    }),
  ).isRequired,
  isLoading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  handleJoinMission: PropTypes.func.isRequired,
  handleLeaveMission: PropTypes.func.isRequired,
};

MissionsView.defaultProps = {
  error: null,
};

export default MissionsView;
