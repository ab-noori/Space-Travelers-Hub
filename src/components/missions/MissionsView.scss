.striped-table {
  width: 100%;
  border-collapse: collapse;
}

.striped-table th {
  white-space: nowrap;
  text-align: left;
  padding: 8px;
  border: 1px solid #dee2e6;
  border-bottom-width: 2px;
}

.striped-table td {
  text-align: left;
  padding: 8px;
  border: 1px solid #dee2e6;
  padding-bottom: 2%;
}

.bold {
  font-weight: bold;
}

.striped-table tr.even-row {
  background-color: #f2f2f2;
}

.striped-table tr.odd-row {
  background-color: #fff;
}

.striped-table tbody tr:hover {
  background-color: #e6e6e6;
}

.button-cell {
  text-align: center;
}

.button-cell button {
  white-space: nowrap;
}

.status-btn {
  height: 1.5rem;
  border: none;
  color: white;
  border-radius: 3px;
}

.action-btn {
  height: 2rem;
  cursor: pointer;
  border-radius: 3px;
}

.active-member {
  background-color: dodgerblue;
}

.not-member {
  background-color: #707070;
}

.leave-mission {
  color: red;
  border-color: red;
}

.join-mission {
  border: 1px solid gray;
}
