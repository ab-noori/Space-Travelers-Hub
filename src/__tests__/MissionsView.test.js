import React from 'react';
import renderer from 'react-test-renderer';
import MissionsView from '../components/missions/MissionsView';

describe('MissionsView', () => {
  it('renders correctly', () => {
    const missions = [
      {
        mission_id: 1,
        mission_name: 'Mission 1',
        description: 'Mission 1 Description',
        reserved: true,
      },
      {
        mission_id: 2,
        mission_name: 'Mission 2',
        description: 'Mission 2 Description',
        reserved: false,
      },
    ];

    const isLoading = false;
    const error = null;
    const handleJoinMission = jest.fn();
    const handleLeaveMission = jest.fn();

    const tree = renderer
      .create(
        <MissionsView
          missions={missions}
          isLoading={isLoading}
          error={error}
          handleJoinMission={handleJoinMission}
          handleLeaveMission={handleLeaveMission}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
