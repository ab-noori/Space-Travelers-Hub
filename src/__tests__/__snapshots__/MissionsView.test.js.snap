// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionsView renders correctly 1`] = `
<div>
  <table
    className="striped-table"
  >
    <thead>
      <tr>
        <th
          className="nowrap"
        >
          Mission Name
        </th>
        <th
          className="nowrap"
        >
          Description
        </th>
        <th
          className="nowrap"
        >
          Status
        </th>
        <th
          className="nowrap"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="even-row"
      >
        <td
          className="bold"
        >
          Mission 1
        </td>
        <td>
          Mission 1 Description
        </td>
        <td
          className="button-cell"
        >
          <button
            className="active-member status-btn"
            type="button"
          >
            Active Member
          </button>
        </td>
        <td
          className="button-cell"
        >
          <button
            className="leave-mission action-btn"
            onClick={[Function]}
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
      <tr
        className="odd-row"
      >
        <td
          className="bold"
        >
          Mission 2
        </td>
        <td>
          Mission 2 Description
        </td>
        <td
          className="button-cell"
        >
          <button
            className="not-member status-btn"
            type="button"
          >
            Not Member
          </button>
        </td>
        <td
          className="button-cell"
        >
          <button
            className="join-mission action-btn"
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
